File: algorithms/sort/insertion-sort/insertion-sort.ts
const swap = (i: number, j: number, array: Array<number>) => {
  const temp = array[j]
  array[j] = array[i]
  array[i] = temp
}
const swap2 = (i: number, j: number, array: Array<number>) => {
  ;[array[i], array[j]] = [array[j], array[i]]
}

export function insertionSort(array: number[]) {
  for (let i = 1; i < array.length; i++) {
    let j = i
    while (j > 0 && array[j] < array[j - 1]) {
      swap2(j, j - 1, array)
      j -= 1
    }
  }
  return array
}

File: algorithms/sort/insertion-sort/insertion-sort.go
package main

import "fmt"

func swap(i, j int, array []int) {
	array[i], array[j] = array[j], array[i]
}

func InsertionSort(array []int) []int {
	for i := 1; i < len(array); i++ {
		j := i
		for j > 0 && array[j] < array[j-1] {
			swap(j, j-1, array)
			j--
		}
	}

	return array
}

func main() {
	xs := []int{3, 2, 100, 1}

	res := InsertionSort(xs)
	fmt.Println(res)
}
